src/gpuarray_array_collectives.c

```diff
@@ -29,15 +29,16 @@ static inline size_t find_total_elems(const GpuArray* array) {
 static inline int check_gpuarrays(int times_src, const GpuArray* src,
                                   int times_dest, const GpuArray* dest,
                                   size_t* count) {
+  gpucontext *ctx = gpudata_context(src->data);
   size_t count_src, count_dest;
   count_src = find_total_elems(src);
   count_dest = find_total_elems(dest);
   if (times_src * count_src != times_dest * count_dest)
-    return GA_VALUE_ERROR;
+    return error_set(ctx->err, GA_VALUE_ERROR, "Size mismatch for transfer");
   if (src->typecode != dest->typecode)
-    return GA_VALUE_ERROR;
+    return error_set(ctx->err, GA_VALUE_ERROR, "Type mismatch");
   if (!GpuArray_ISALIGNED(src) || !GpuArray_CHKFLAGS(dest, GA_BEHAVED))
-    return GA_UNALIGNED_ERROR;
+    return error_set(ctx->err, GA_UNALIGNED_ERROR, "Misbehaved arrays");
```

Comment:
unaligned or misbehaved.
I don't remember what is misbehaved myself, so we can't expect user to also remember. I think you should split that into 2 different error and give a better error for the misbehaved version.